### PR TITLE
Adds Google Analytics tracking (fix #56)

### LINKED ISSUE #56: add google analytics code
possible implementation approach from [http://guides.emberjs.com/](http://guides.emberjs.com/v1.12.0/cookbook/helpers_and_components/adding_google_analytics_tracking/)

### PR DESCRIPTION
**Pull request by jamemackson**
@IanVS feel free to review and merge

### CODE REVIEW

File: client/app/router.js [hunk 1]
```diff
@@ -21,7 +37,7 @@ export default Router.map(function () {
     this.route('bio');
     this.route('statement');
     this.route('press');
-    this.route('image', { path: '/image/:slug' });
+    this.route('image', { path: '/art/:slug' });
     this.route('edit', function () {
```
Comment (IanVS) [inline client/app/router.js]: Is this related to the analytics, or could it be in a separate PR?

Comment (jamemackson) [inline client/app/router.js]: it is related to analytics in that having `art` in the url will be better for analytics from an SEO perspective than having `image` there so I slipped it in here.

File: client/Brocfile.js [hunk 2]
```diff
@@ -1,11 +1,23 @@
-/* global require, module */
-
+/* global require, module, process */
+var env = process.env.EMBER_ENV;
```
Comment (IanVS) [inline client/Brocfile.js]: Is there a way to get the ember environment some other way than by an environment variable? That seems sub-optimal.

Comment (jamemackson) [inline client/Brocfile.js]: I would guess there probably is a better way, I was just following the examples here: https://www.npmjs.com/package/ember-cli-inline-content .  This approach seems to work for now so I figured we can circle back if we don't have a better mechanism to use now.  I'm wide open to suggestion however.

Comment (jamemackson) [inline client/Brocfile.js]: Looking back at this code again now (in the light of day) I think we could wait to initialize the `config` variable after we have instantiated the `app` variable and then declare `config` var like so
```
var EmberApp = require('ember-cli/lib/broccoli/ember-app');
var app = new EmberApp({
  hinting: false
});
var config = require('./config/environment')(app.env);
```
The only longer term downside to this approach is that we wouldn't have access to the config when initializing `app` but since we are not using that variable in it's instantiation presently I don't see that as a problem now.

Comment (IanVS) [inline client/Brocfile.js]: The method suggested by  [ember-cli-google-analytics] (https://github.com/pgrippi/ember-cli-google-analytics#configuration) seems more elegant to me.  Not that we need to use that addon, but the idea of adding a property to ENV seems better than a environment var set in the OS.

Comment (jamemackson) [inline client/Brocfile.js]: yes that does look like a good plugin.  We are currently adding the [GA property id to ENV](https://github.com/hairbrain/gallery13/blob/fix_56/client/config/environment.js#L7) also.  We are only using the environment variable when loading the config object to know what environment is being run.  

I wonder if the current approach gives us more flexibility to control how and when the events are fired compared to ember-cli-google-analytics.  Pageview event tracking is typically the baseline analytics that gets setup and then adding event tracking is typically the next thing that is asked for.  It seems to me that the current approach will give us more flexibility long term w/o having to submit PR's to ember-cli-google-analytics to add this logic.  I may be wrong, I haven't looked into the details of that plugin beyond a quick read of their README...

File: client/config/environment.js [hunk 3]
```diff
@@ -37,17 +38,15 @@ module.exports = function(environment) {
     ENV.contentSecurityPolicy = {
       'default-src': "'none'",
       // add the domain here like analytics.google.com or data: or 'self'
-      'script-src':  "'self' 'unsafe-eval' 'https://www.google.com' ''",
+      'script-src':  "'self' 'unsafe-eval' 'unsafe-inline' www.google.com www.google-analytics.com",
```
Comment (IanVS) [inline client/config/environment.js]: I'm not sure this is a good idea from a security perspective. It basically opens us up to XSS attacks.  Check out "Inline Code Considered Harmful"  in http://www.html5rocks.com/en/tutorials/security/content-security-policy/

Comment (IanVS) [inline client/config/environment.js]: It seems like the script is not inline, you've put it in ga.js, so we shouldn't need 'unsafe-inline'  here, should we?

File: client/config/environment.js [hunk 4]
```diff
@@ -65,10 +64,10 @@ module.exports = function(environment) {
     ENV.contentSecurityPolicy = {
       'default-src': "'none'",
       // add the domain here like analytics.google.com or data: or 'self'
-      'script-src':  "'self' 'unsafe-eval' 'https://www.google.com' ''",
+      'script-src':  "'self' 'unsafe-eval' 'unsafe-inline' 'https://www.google.com' 'www.google-analytics.com'",
```
Comment (IanVS) [inline client/config/environment.js]: Can we remove 'unsafe-inline' here?

Comment (jamemackson) [inline client/config/environment.js]: Pretty sure we still need that but I can test it out again when I get back to my desk

Comment (IanVS) [inline client/config/environment.js]: You're right, from the addon docs:

> During the build process, this will render the contents of those files directly inline with <script>

Is this the best strategy if it opens up XSS attacks?  I believe it is recommended not to inline JS in general, right?

**Comment by IanVS** [comment]
Some good notes from Google about analytics on SPAs: https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications

File: client/config/environment.js (hunk 4 above)
Comment (jamemackson) [inline client/config/environment.js]: I think to get away from the inline script approach we would have to model our solution in a similar fashion to how their are loading the live reload script with ember-cli [ember-cli-inject-live-reload](https://github.com/rwjblue/ember-cli-inject-live-reload).  

I'm thinking we should make create an issue for this and circle back later on this.  Maybe give a PR to [ember-cli-google-analytics] (https://github.com/pgrippi/ember-cli-google-analytics#configuration) at that time and migrate to using that.

**Comment by IanVS** [comment]
Is this ready to merge?

File: client/config/environment.js (hunk 3 above)
Comment (jamemackson) [inline client/config/environment.js]: I just tested it out without having `'unsave-inline'` specified and did not get any security policy warnings so I'm thinking we can move forward with these changes with out it.  We can always add it back if we determine there are issues once these changes are deployed.

Comment (jamemackson) [inline client/config/environment.js]: ok I take that back.  We do need it.  I was forgotten that I had basically disabled the real ga code when not in production, when I re-enabled it locally I did again get the policy errors and the pageview code did not fire.  I have added it back in again and verified the ga logic is working in the realtime analytics dashboard.

image

**Comment by jamemackson** [comment]
so pending a final review, I think this is ready to be merged.  :+1:

**Comment by IanVS** [comment]
Want to go ahead and squash commits?

File: client/config/environment.js [hunk 5]
```diff
@@ -4,12 +4,14 @@ var devHost = 'http://localhost:1337';
...
 var devS3 = 'https://gallery13-dev.s3.amazonaws.com';
 var prodS3 = 'https://gallery13.s3.amazonaws.com';
+var gaId = 'UA-12392326-1';
 
 module.exports = function(environment) {
   var ENV = {
     modulePrefix: 'client',
     podModulePrefix: 'client/pods',
     environment: environment,
+    isProduction: (this.environment === 'production'),
```
Comment (IanVS) [inline client/config/environment.js]: I'm still not convinced this is the best solution, it feels a bit like a workaround.  Couldn't you set the Google settings directly within https://github.com/hairbrain/gallery13/blob/fix_56/client/config/environment.js#L31 and https://github.com/hairbrain/gallery13/blob/fix_56/client/config/environment.js#L63?

Then, instead of testing for environment within the brocfile, you could just look at the google analytics settings?  As it is now, there seem to be a lot of duplicate checks.

Comment (jamemackson) [inline client/config/environment.js]: Yes we could further simplify those settings as you suggest.  

I had added the `isProduction` property here to reduce the number of times I was hard coding `environment === 'production'`.  Ultimately we just want the ga code to run if we are in production or if we have the debugging enabled based on `ENV.debugGoogleAnalytics` 

So yes, I we could just change those places to look at if `ENV.googleAnalyticsId` is set though and that might simplify things...

Comment (IanVS) [inline client/config/environment.js]: To be clear, what I'm proposing is:

In config/environment.js:
```js
  const DEBUG_GA = false;
  const GA_ID= 'UA-12392326-1';

  if (environment === 'development') {
    ...
    if (DEBUG_GA) {
      ENV.googleAnalyticsId = GA_ID;
    }
  }
  if (environment === 'test') {
    ...
  }
  if (environment === 'production') {
    ...
    ENV.googleAnalyticsId = GA_ID;
  }
```

And in brocfile:
```
var config = require('./config/environment')(env);

if (config.googleAnalyticsId) {
  app.options.inlineContent = {
    'google-analytics' : {
      file: './ga.js',
        postProcess: function(content) {
        return content.replace(/\{\{GOOGLE_ANALYTICS_ID\}\}/g, config.googleAnalyticsId);
      }
    }
  }
}

```

Comment (jamemackson) [inline client/config/environment.js]: cool, that looks fine by me.

Comment (jamemackson) [inline client/config/environment.js]: I can make those changes when I get back home and will attempt to squash my commits :) then we can merge this one in.

**Comment by IanVS** [comment]
LGTM

**Comment by jamemackson** [comment]
alrighty then.  finally squashed!  :+1:
